### ISSUE TITLE
Suggestions

### ISSUE THREAD

**Issue by dev-101** [CIDRAM/CIDRAM#3]
Hi Maik, here's few:

**1)** I think bogon (`**edit/update:**` it is not caused by bogon, wp_cron is still being blocked from other rules/ranges) option should be set to false by default, because of this:

> ID: 1
> Script Version: CIDRAM v0.1.2
> Date/Time: Wed, 16 Mar 2016 12:44:17 +0100
> IP Address: 192.168.0.1 (modified, left for illustration purposes only)
> Query: doing_wp_cron=1458128657.1168138980865478515625
> Referrer: 
> Signatures Count: 1
> Signatures Reference: 192.168.0.0/16 (modified, left for illustration purposes only)
> User Agent: WordPress/4.4.2

Now, we could at least add bypass IPs to the list, so that they are excluded from the above rules.
Additionally, bogon description here sounds... well, martian/strange. If its main purpose is to block local range, shouldn't it be called more user-friendly, **block_local** perhaps? :)

**2)** outgen.inc (& others .inc) should be renamed to outgen.php, let's depart from the terrible practice of the past, and name files with proper extensions :wink:

If .inc is so desperately needed, we could keep it, but append .php finally in the end.
My code editor & me likes it.

**3)** I might be missing something, but I don't see here Why? section :confused:
It should be added as short as possible, but informative (e.g. "ONLINE S.A.S.")

> ID: 2
> Script Version: CIDRAM v0.1.2
> Date/Time: Wed, 16 Mar 2016 13:11:42 +0100
> IP Address: 212.129.56.230
> Query: 
> Referrer: 
> Signatures Count: 1
> Signatures Reference: 212.129.0.0/18
> User Agent: PHP/5.3.56

**4)** Surely I am still catching up :) but is there a whitelist (**ipwldb.csv**) somewhere? Or do I need to use my good old wrapper? `**edit/update:**` already did, had to copy ipwldb.csv and use it, to prevent wp_cron from not running.

**5)** Would be nice to have option in config.ini if we wish 403 or 503 headers. Just throwing my idea here first, should be easy to implement later.

Thanks

**Comment by Maikuolan** [comment]
Hi dev-101,

Thanks for the suggestions! :-)

I'm just about to get some sleep for the night, but, I felt I should quickly reply here first, to let you know that I've seen your message here (and so that it won't look like I'm ignoring it).

I'll reply here again, more properly and more thoroughly, in the morning, after I've woken up.

**Comment by dev-101** [comment]
Ok, thanks for response!
Will try to find some time and do some actual work :)

**Comment by Maikuolan** [comment]
> Additionally, bogon description here sounds... well, martian/strange. If its main purpose is to block local range, shouldn't it be called more user-friendly, **block_local** perhaps? :)

More-so, the main purposes of including the "bogon" CIDRs is to block requests originating from any IPs that we wouldn't normally expect to receive traffic from due to being non-routable within the public IP address space (so, "bogon" IPs would be a combination of local range IPs, unassigned IPs and reserved IPs). Strictly speaking, these IPs aren't necessarily "malicious", but rather, the logic for blocking them is that, if we're not expecting traffic from these IPs, traffic from these IPs could be considered weird or unusual, and thus, there is a question of why they're sending requests, and by extension, the possibility that they _could_ be malicious.

Because of these in the IPv4 signature file:

```
10.0.0.0/8 Whitelist
127.0.0.0/8 Whitelist
192.168.0.0/16 Whitelist
```

..and this in the IPv6 signature file:

```
0::/8 Whitelist
```

..generally, the most commonly used local IPs shouldn't normally be blocked by the bogon signatures, but, because local IPs can, technically speaking (and with a little messing around with router and/or DHCP settings), be assigned to anything within the entire IPv4/IPv6 address space (and don't necessarily need to conform to whatever personal routers would normally assign by default), and because there are some particular brands of routers that assign local IPs weirdly, in ways not considered "normal", I still would recommend to any users that expect traffic to their websites from the local network to disable the directive for bogon signatures.

Because bogons _can_ extend beyond what we'd consider the "normal" ranges for local addresses, I do think that "bogons" is the most appropriate term to use to describe these CIDRs collectively, but, it may perhaps be a good idea to have them disabled by default (seeing as most users probably won't need them, but they could nonetheless serve as an additional layer of protection for those that want to be extra cautious against any possibilities for unwanted IPs connecting to their sites).

I will concede, though, that "local IPs" is a far more recognisable and intelligible description than "bogon IPs", due to average users likely not knowing the proper definition of the term "bogon" in the context of networking (whereas most average users will know what a "local IP" is), so, perhaps a re-wording or better description of these within the documentation and/or signature files may be appropriate (I can see to this being done). :-P

> **2)** outgen.inc (& others .inc) should be renamed to outgen.php, let's depart from the terrible practice of the past, and name files with proper extensions :wink:

I've always felt that using only the `PHP` file extension makes more logical sense than using multiple other extensions (such as `INC`), but, the main reason why I've been using `INC` for includes is due to problems I've experienced firsthand as well as heard from others experiencing in the past whereby auto-deployment and auto-update tools can/will sometimes have problems with creating/updating files with the `PHP` extension due to overzealous server-side security solutions (firewalls, anti-virus, etc) automatically blocking/deleting any/all files with the `PHP` extension whenever they're created/updated via any means other than FTP/SSH/etc (this is usually done with good intentions, such as to prevent the installation of backdoors or the exploitation of illegally installed shell scripts whereby these backdoors/shells will try to install their own unwanted and usually malicious PHP files onto the server and etc, but unfortunately, when auto-deployment and auto-update tools are wanting to create/update/install PHP files, because this will appear to be the same behaviour as far as these firewalls and anti-virus solutions are concerned, these auto-deployment and auto-update tools generally will be thwarted in the same manner that backdoors and shells would be).

That all said, though.. Other than Packagist, I'm not aware of there being any auto-deployment or auto-update tools that will deal with CIDRAM, and I haven't coded any auto-update functionality into CIDRAM itself, so, the above point may be somewhat negligible in this instance, for this project.

I'm not strictly against renaming all the `INC` files to `PHP` files, but, due to the above concerns, I wouldn't mind hearing from others before doing this, to know what others think about this idea. I'll ask around, to see if I can get anyone else to give their input here. :-)

> If .inc is so desperately needed, we could keep it, but append .php finally in the end.
> My code editor & me likes it.

Not desperately needed; Just needed as a way to avoid problems with auto-deployment in some situations, so, we _may_ be able to get rid of the `INC` extension. Renaming files to `%filename%.inc.php` probably wouldn't help as a compromise between the need and the change anyhow, due to that these files would then still have the `PHP` extension (and therefore, in theory, would still pose this problem for auto-deployment/auto-update tools).

> **3)** I might be missing something, but I don't see here Why? section :confused:
> It should be added as short as possible, but informative (e.g. "ONLINE S.A.S.")

Nope, you're not missing anything there. I'll see to getting this fixed soon, within one of the coming versions/releases. I agree, something should be added here. :-)

(I'll likely implement this in a slightly different way to ZB Block, though, because I'd like to keep management/maintenance of the signatures as absolutely simple and easy as possible, and the way that each ZB Block signature has their own description by necessity, has the potential for increasing maintenance complexity which could eventually lead to maintenance problems in the future; I'm going to try writing code that will allow custom descriptions to be optional, defaulting to something more generic; similar to how the reason presented to the user works; this is already different to ZB Block as is anyhow, due to that ZB Block presents the same reason to the user as to the logfiles, and CIDRAM already presents a reason for being blocked to the user, but just isn't logging that reason yet; also, in that CIDRAM specialises in dealing with CIDRs, whereas ZB Block can block requests for other reasons, tracking/debugging signatures in CIDRAM should be much easier, due to the "Signature Reference" label both presented to the user and included in the logfile, which can be used to quickly search for the relevant signatures within the signature files; though, a friendlier description of the reason blocked to be logged upon requests being blocked, could be beneficial).

> **4)** Surely I am still catching up :) but is there a whitelist (**ipwldb.csv**) somewhere? Or do I need to use my good old wrapper? `**edit/update:**` already did, had to copy ipwldb.csv and use it, to prevent wp_cron from not running.

Don't need a separate whitelist file for CIDRAM. :-)

Because CIDRAM only blocks requests on the basis of the originating IPs of requests, there's no mechanism in CIDRAM for having a user/client being accidentally blocked for any reason other than their IP address, so, simply using the whitelisting mechanism available to CIDRAM for whitelisting CIDRs should be sufficient.

To do this (to whitelist a CIDR), all you need to do in include a CIDR in your custom signatures file with the "Whitelist" suffix added to it, and it should be whitelisted.

Something like:

```
1.2.3.4/32 Whitelist
```

..will suffice (the above CIDR used only as an example, of course).

This is also described within the custom signature files themselves, with better detail describing how this mechanism works.

> **5)** Would be nice to have option in config.ini if we wish 403 or 503 headers. Just throwing my idea here first, should be easy to implement later.

Hm, true. Maybe what I'll is modify the `forbid_on_block` directive, so that it can accept the currently accepted values (true or false, for a 200 or 403 response), or accept a different, specified header, if desired. That could be a way to implement this feature with minimal impact to the default configuration and to minimise the extent of changes. Anyhow, I'll have a play around with this idea, to see what I can come up with.

> Ok, thanks for response!
> Will try to find some time and do some actual work :)

No problem! And thanks for the suggestions. :-)

**Comment by dev-101** [comment]
Hi Maikuolan, here's to someone who likes to write long replies :beer:
I will be short, on the other hand :)

( reply **2** )

I think you are overthinking this one, I have never seen the problem you described, unless it was improperly configured or generic server setup with messed-up users and owners. Think WordPress, hundreds/thousands of .php files and almost weekly updates from within the script itself (yes, even autoupdates), that works like a clockwork. Don't worry that much, and move to .php already, as it is the only logical thing.

( reply **4** )

This should be part of the main README in a brief form (because most read only the main project page at first) and docs. I have read dat files, but later simply forgot (there is too much text information in there, and my focus was on blocking rules :) ).

Still, I just feel too lazy now to maintain another whitelist file, so will keep my wrapper for now. Maybe in the future, if some merger happens, who knows... :grey_question:

---

Thanks again for your comprehensive reply :+1:

**Comment by Maikuolan** [comment]
(**2**)

True, true. This same issue would probably apply to some of my other projects, too (like phpMussel, for example), where INC files are used for the project includes.

(**4**)

Yeah, I think I'll do that. Makes sense. :+1:

**Comment by Maikuolan** [comment]
I spoken with a few others, and most of them seem to be for the idea of changing from `INC` to `PHP`, so, I'll get started on this shortly. :+1:

**Comment by Maikuolan** [comment]
(**1**) and (**2**) are now implemented. The others should be implemented soon. :+1:

**Comment by dev-101** [comment]
Hi, just thought of another potential issue:

**6)** `ipv4_custom.dat, ipv6_custom.dat`

I think that above files should not be part of the package initially, because they will be overwritten with every "blind" upgrade. Instead, the content should be moved into manual or "template" equivalent files ipv4_custom.dat**.txt**, ipv6_custom.dat**.txt** (.txt should be removed by user later).

**Comment by DanielRuf** [comment]
Definitely makes sense. Or ipv4_custom.example.dat or something else. Anyway, it is always recommended to make a full backup before updating as also the config file could be overwritten.

But when we switch to Packagist, Composer and so on, these templates should be part of another optional package or optionally available or dynamically created by the main script if they do not exist.

Files in a vendor package should not be changed if possible. Or always make a full backup. This is also recommended by big CMS (files like .htaccess and so on).

**Comment by dev-101** [comment]
True, forgot about config.ini :+1:

Maybe basic `installer` (or `config generator` more accurately) could be added which will compile config.ini from config.template.ini, but it should be renamed as well.

Included .htaccess files are great, as long as assumption about Apache compatible server is true :)

**Comment by DanielRuf** [comment]
Generating these would add another extra step and need more code. 

Generally I am for doing a full backup and reapplying the settings.

Files can be easily compared with diff tools like WinMerge and other tools on the CLI (using SSH) and against generating more files which could also introduce new bugs and problems like insufficient chmod / chown rights and so on.

@Maikuolan How do you think about this?

**Comment by dev-101** [comment]
I had a simple stand-alone install.php file in mind (could be manually deleted as well) that would just present few config options available and generate config.ini file afterwards. But, for security reasons install.php should be accessible only by admin, so that may be tricky with fresh upgrades and web access (since, if standard directory name is used and someone pokes around...), also checking if config.ini already exists in the first place, to avoid either accidental or intentional overwriting.

Of course, manual approach with renaming config sample is also fine, more secure and advanced.

**Comment by dev-101** [comment]
From docs:

> I hope to streamline this process by making an installer at some point in the not too distant future

**Comment by DanielRuf** [comment]
The documentation is a changed copy of the phpMussel documentation ;-)

**Comment by Maikuolan** [comment]
All of the above, true, and all good points. XD

Most of the CIDRAM documentation I'd copied directly from phpMussel, and modified as per necessary (because the two packages have extremely similar structures, a significant portion of the documentation could apply equally to either of them, requiring very little modification; all necessary modifications have been made, of course, but much of it remains nearly verbatim). This also makes i18n/translation work much, much easier (because if the documentation of both packages is similar and very comparable, any translations done for one package can be easily replicated for the other package).

The point about wanting to make an installer at some point in the not too distant future, is true, and could be said for both packages: This is something that I would like to do (eventually, whenever that may be), and it's something that we could look into doing. It has been a desire since day one, but, also, it's something that I've perpetually postponed and haven't yet started working towards, despite being a desire since day one, because it's something that I think should be very carefully planned out beforehand, before any work is done towards making it (due to numerous considerations, such as security, the potential to introduce new complications or vulnerabilities, and we would need to figure out and decide the best way to proceed as so that it doesn't become more trouble than it's worth for users and for us).

I would be interested in discussing it to plan it further, but I think that implementing an installer would likely be a reasonably large job, so, it might be better to create a separate, dedicated issue for discussing this (so that the various other suggestions made don't get drowned out by any potential dominance of the installer suggestion).

This could also possibly tie into future plans for a package front-end (to allow changing configuration directly via the browser, or, for in the case of phpMussel, as a complete replacement for the current "Browser Controls" handler/sub-system), which is also something that has been desired for a while now (but postponed for similar reasons).

I'm also a big fan of minimising the amount of work required for users to get CIDRAM (and phpMussel, too) working correctly on their systems, so, a method that reduces the amount of work required by them is always good (and so, it would be good to avoid unnecessary additional steps, wherever possible).

Auto-compiling the configuration is definitely a good idea. Ideally, a more secure and more malleable way of storing the configuration than is currently the case (whereby we're using the `config.ini` file) would be good, too (like how Matthias suggested the ".htsecurity" idea a while back), but this is also something to discuss.

About to get some sleep for the night; Will reply again in the morning.

**Comment by Maikuolan** [comment]
> Definitely makes sense. Or ipv4_custom.example.dat or something else. Anyway, it is always recommended to make a full backup before updating as also the config file could be overwritten.
> 
> But when we switch to Packagist, Composer and so on, these templates should be part of another optional package or optionally available or dynamically created by the main script if they do not exist.
> 
> Files in a vendor package should not be changed if possible. Or always make a full backup. This is also recommended by big CMS (files like .htaccess and so on).

Agreed. I particularly like the idea of possibly including example files or of auto-generating files; I think these ideas would be preferable above having a secondary package for these, because auto-generating rather than keeping them as a secondary package might be easier for users to deal with.

Alternatively, maybe we could have it that the CIDRAM searches the vault for files of a particular naming convention (such as `ipv4_*.dat`, `ipv6_*.dat`, etc) within the package and processes them as signature files, instead of as is the currently the case whereby CIDRAM is fed which files to process as signature files via predefined arrays? This could potentially help make things easier for users that want to use multiple additional signature files (such as for users making use of **[MacMathan's country blocks](http://macmathan.info/zbblock-range-blocks)**, which are now offered both in ZB Block signature format and CIDRAM signature format); It's already possible currently to add additional signature files to CIDRAM, but not without modifying the predefined arrays in the `functions.php` file so that CIDRAM can know which files to process as signature files. If we changed this so that CIDRAM could determine which files to use as signature files automatically based upon naming schema or upon some similar thing, then, there could be instructions added to the current signature files to simply rename them based on the naming schema or duplicate them into multiple files based on the naming schema.

Thoughts?

**Comment by DanielRuf** [comment]
This is a good idea (dynamic array) which can be solved with `glob` http://php.net/manual/en/function.glob.php

So this array would be automatically generated (and properly sorted).

**Comment by dev-101** [comment]
Personally, I don't like idea of scanning ipv_ files with each request performance wise. Despite the caching and all, it will still take time. Instead, simple reserved custom entry in config that will accept comma separated values could be enough.

**Comment by dev-101** [comment]
**7)** This is optional, but I would really like to distinct comment lines apache- or other conf files alike, starting those lines with #. This could potentially resolve current requirement of linux-type line breaks and improve compatibility in signatures. Also, it would add benefit of highlighting in editors, because now we cannot distinct commentary from the actual rules in no way

**Comment by Maikuolan** [comment]
Hm, true. Scanning shouldn't really take too long, but reserved, custom entries would still be quicker, though (and it's good to be as fast as possible with these things).

In any case.. Changing the way we handle custom signature files is something to think about (so that we can avoid needing to edit the core when names of files are changed and/or when files are added).

**7)** We could easily added some "#" or similarly styled comment prefixes to the signature files if we wanted to, to make them easier to read and to highlight for editors; I'm not against this idea. I don't really like the idea of _forcing_ users to _conform_ to any particular comment style, though, because currently, anything not recognised as a signature by CIDRAM is just ignored entirely during execution, and so, to determine whether any non-signature data is a comment that correctly conforms to some particular style would require additional processing (ie, we couldn't continue to ignore any data during execution) to go through all that non-signature data to check through it all (which, although helpful as a way to tell users when they've done something wrong in the signature files, would slow down processing time and wouldn't add any benefit during execution). So, basically, I'm happy to change the way we write comments by default, but I don't think we should force it upon anyone.

**Comment by dev-101** [comment]
That 'silent' rejection of improperly defined rules is exactly most disconcerting for me. If we add something to the rule set, we cannot be sure if it will work or not, until [x] hits us again and logging is enabled.

By contrast, in zbblock when you add a badly defined rule (e.g. forgotten quote or semicolon) you will trigger a PHP error and the site will be inaccessible, so you can be "damn sure" that you messed it up or set it properly. The added complication of different OS-es line breaks does not help to the issue, either.

**Comment by dev-101** [comment]
I think we could issue a PHP notice (or a warning) if particular rule was rejected (and add option in config.ini to turn it on or off), which would be a very useful indicator. But, this approach asks for clear comment hashtags, to avoid false reporting.

**Comment by DanielRuf** [comment]
Different linebreaks should not be an issue.

We could just throw a custom exception if something is not ok, notices, warnings, fatal errors and so on are internal errors.

Theoretically rules can be easily tested with curl eg as unittest.

**Comment by DanielRuf** [comment]
There is just `trigger_error` and the flags `E_USER_*` where \* can be ERROR, NOTICE, WARNING and DEPRECATED. Custom exceptions are much better imho.

**Comment by dev-101** [comment]
Maybe I was not clear enough, but that is what I had in mind.

**Comment by Maikuolan** [comment]
Just synced a new commit.

> CIDRAM signature files can now support both Unix and Windows linebreaks ("\n" or "\r\n") without breaking signatures (this was done by adding a check for the presence of "\r" in signature files, and if found, stripping them out prior to processing). Partially implemented a "why blocked" field for logging and page output (although there's still more work to be done). Corrected placement of some incorrectly placed comments.

About to disconnect for a few hours, but I'll reply again when I next come back online.

**Comment by Maikuolan** [comment]
..How about an external validator/fixer? That shouldn't be too hard to write up, and we could use it to check over signature files whenever they're modified as to ensure that the signature files conform to our rules; That way, we'd know whether we'd botched anything up, without needing to send warnings and notices to the end user/client during the normal execution phase.

It could be included either as part of the main repository or as a separate repository that could be accessed by any users that want to modify their signature files or use custom signatures (perhaps `CIDRAM-Validator` or `CIDRAM-Fixer`).

The additional processing required of checking over everything wouldn't matter much anyhow, too, if we used an external validator/fixer (because anything done external/separate from the main execution phase won't be noticed by the end user/client).

**Comment by DanielRuf** [comment]
Makes much sense, an extra repository would be very good. Could use the plugin architecture or be independent of the main repository. A check should be just once when the files which need to be checked have changed. Could be easily added with Composer then.

**Comment by Maikuolan** [comment]
That it could. :-)

And I'm thinking the same, too, that this should probably only be checked when changes are made, rather than on every single execution.

**Comment by Maikuolan** [comment]
All signature files have been modified to conform with **(7)** (although none of the core code has been modified yet in regards to this suggestion).

**(6)** has been implemented (for signature files; custom signature files have had a `.RenameMe` appended to them); Configuration changes not yet implemented.

**Comment by Maikuolan** [comment]
Configuration file has been renamed to `config.ini.RenameMe` (although I haven't yet started any work on an installer or a browser-based configuration guide).

**(5)** Ability to return 503 headers has been implemented (via extending the "forbid_on_block" directive).

With that.. I think the key ideas behind each of these suggestions has now been implemented (although some of the README translations for **(4)** are still pending, but, that won't take too long to do). :+1:

**Comment by Maikuolan** [comment]
(_Work-In-Progress_ label removed and _Implemented_ tag added; Leaving the issue open though, due to work on an installer having not yet been started and due to the possibility for follow-up discussion on the suggestions made).

**Comment by dev-101** [comment]
Thanks Maik, sorry I was not actively participating lately, but you have done great work so far :+1:

**Comment by Maikuolan** [comment]
Not a problem, and cheers. :+1:

**Comment by Maikuolan** [comment]
_Mentioning the front-end installer idea in a new issue so that we can close this old issue._
